### PR TITLE
docs/API.md: mention /upload in API doc

### PR DESCRIPTION
**Pull request by rogpeppe**
We update the API docs to describe the new upload endpoint,
and also add a new endpoint (as yet unimplemented) to
enable a client to find out in advance if it has permission
to modify a charm.

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2067/

File: docs/API.md [hunk 1]
```diff
@@ -2376,3 +2398,91 @@ Example: `GET changes/published?limit=10&start=31-07-2014`
...
+	// be uploaded.
+	MaxPartSize int64
+
+	// MaxParts holds the maximum number of parts.
+	MaxParts int
+}
+```
+
+Once an upload has been started, all its parts should be uploaded
+(they do not have to be uploaded in order), and then finished
```
Comment (bac) [inline docs/API.md:2442]: finished with a put

File: docs/API.md [hunk 2]
```diff
@@ -2376,3 +2398,91 @@ Example: `GET changes/published?limit=10&start=31-07-2014`
...
+will be discarded. The response holds a JSON
+object containing information about the upload:
+
+```go
+type NewUploadResponse struct {
+	// UploadId holds the id of the upload.
+	UploadId string
+
+	// Expires holds when the upload id expires.
+	Expires time.Time
```
Comment (bac) [inline docs/API.md]: As discussed on IRC perhaps it is worth mentioning that the user-supplied expiration time must be <= 24h from the upload start time.

Are there restrictions on the format or timezone used? Does it have to be Zulu?

File: docs/API.md [hunk 3]
```diff
@@ -2376,3 +2398,91 @@ Example: `GET changes/published?limit=10&start=31-07-2014`
...
+This endpoint uploads a single part with the given part number *part*
+to the upload with the given *uploadid*. The hash parameter must specify
+the SHA384 hash of the uploaded part in hexadecimal format. The data
+is read from the request body. The request must specify the size of the
+data in its Content-Length header.
+
+#### PUT /upload/*uploadid*
+
+This endpoint completes an upload. The body should contain a JSON object
+holding the hashes of all the parts that have been uploaded. One this
```
Comment (bac) [inline docs/API.md]: Typo: Once

**Review by bac — APPROVED** [review]
Looks good with one suggestion and a few typo corrections.

**Review by mhilton — APPROVED** [review]
:+1:

File: docs/API.md (hunk 2 above)
Comment (rogpeppe) [reply docs/API.md]: The format is RFC3339. I've mentioned the expiration time limit and format.

File: docs/API.md (hunk 1 above)
Comment (rogpeppe) [reply docs/API.md:2442]: done

File: docs/API.md (hunk 3 above)
Comment (rogpeppe) [reply docs/API.md]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2069/

**Pull request merged**